### PR TITLE
roachtest: add multi-cluster support and introduce asymmetrical CPU test

### PR DESCRIPTION
**Pull request by ajstorm**
## Summary

This PR introduces comprehensive multi-cluster support to the roachtest framework and implements an asymmetrical CPU test for cluster-to-cluster replication.

Previously, roachtest only supported single-cluster tests, which limited our ability to validate cluster-to-cluster replication under realistic conditions where source and destination clusters have different hardware configurations.

## Changes

**Multi-cluster infrastructure:**
- Added MultiCluster interface and implementation for managing multiple independent clusters within a single test
- Enhanced TestSpec with multi-cluster configuration support including named clusters (e.g., "source", "destination", "workload")  
- Updated test registry to handle multi-cluster test registration and provisioning
- Improved test runner to provision and manage multiple clusters concurrently

**Asymmetrical CPU test:**
- Implemented c2c/asymmetrical-cpu test that validates cluster-to-cluster replication performance when destination cluster has fewer CPU resources than source
- Source cluster: 3 nodes with 16 CPUs each (n2-standard-16)
- Destination cluster: 3 nodes with 8 CPUs each (n2-standard-8)
- Workload cluster: 1 node for generating continuous load
- Runs for 30 minutes monitoring end-to-end replication latency

This enables testing of real-world scenarios including heterogeneous hardware configurations, disaster recovery setups, and multi-region deployments with different resource allocations.

Release note: None

Epic: None

### CODE REVIEW

**Comment by ajstorm** [comment]
@herkolategan or @srosenberg, any chance I could get some 👀 on this PR?  Even a quick glance to see if this is something that you'll at all consider would be very helpful.

**Comment by herkolategan** [comment]
@ajstorm No problem; will have a look by tomorrow.

First thought that comes to mind is that the issue this PR solves is more an issue of having different machine types, rather than different clusters. But I remember there being some issue with making a non-homogeneous cluster cc @DarrylWong.

**Comment by DarrylWong** [comment]
> But I remember there being some issue with making a non-homogeneous cluster 

There shouldn't be an issue, `c.WorkloadNode` is implemented as a different sized machine in the same roachprod cluster. The API to do this isn't exposed to roachprod, which a few people have asked for which is what you might be thinking of. 

> is more an issue of having different machine types, rather than different clusters

To me they are two distinct issues that we both want to solve. PCR/LDR tests do want to have different distinct clusters to get around various roachprod limitations, e.g. Needing to start the src before the cluster because node 1 is the init node, needing to do constant node arithmetic to find node 1 on the dst cluster. 

We've also had one ask to be able to spin up a single cockroach cluster with different machine types, which wouldn't be solvable with multiple clusters.

It does feel to me that the former is a much more niche use case than having multi clusters i.e. many of the reasons you'd want a non-homogeneous cluster could be better solved by just creating two roachprod clusters with certs setup. The aforementioned workload node as one example where I implemented it as a non-homogenenous cluster for simplicity reasons only.

------

I've only skimmed the PR, but it does make me a bit hesitant wrt maintainability due to the branching `IsMultiCluster()` checks we will have to start thinking about. For example, how does this interact with cluster reuse? It seems we will never be able to reuse clusters between `MultiCluster` and non `MultiCluster` tests.

What if we run subsequent `MultiCluster` tests? Will it attempt to reuse the first cluster and never enter the branch to create the second cluster?

That being said, I'm not sure there is a clean way to implement this in the current test framework. How to support multiple clusters is something i've been thinking about recently as well, especially in the context of modular testing if we are to create a new framework. So far I'm thinking that it'd be nice if tests could create clusters as needed instead of needing to declare them at test spec time. Of course this also has it's own problems with managing CPU quota and cluster reuse, which I don't have a good solution for yet :)

**Comment by herkolategan** [comment]
> want to have different distinct clusters to get around various roachprod limitations, e.g. Needing to start the src before the cluster because node 1 is the init node, needing to do constant node arithmetic to find node 1 on the dst cluster.

Ah yes this was the issue with supporting different machine types in the same cluster, it doesn't really solve those edge cases. And a non-homogeneous cluster ends up not being reusable in most cases either.

> That being said, I'm not sure there is a clean way to implement this in the current test framework.

I share the same concern, and I'm not sure we can get a clean abstraction for this without a major refactor.

**Comment by ajstorm** [comment]
Thanks for having a look @DarrylWong and @herkolategan.  Where does this leave us?  We have a desire to test this functionality and no means to do so without either (a) a partial solution like the one propose here, or (b) a significant refactor.  I have to presume that (b) isn't on your team's roadmap, so does it make sense to proceed with a partial solution for the time being?

**Comment by herkolategan** [comment]
@ajstorm The need for this makes sense. After having a look at the code, I have a proposal to reduce the footprint a bit. Since a `roachtest` always needs at least one cluster; instead of phrasing it as multi-cluster tests, we could rephrase it as adding a spec option for supplementary clusters. Hence, a `roachtest` will always have a "primary" cluster (which is the same as we have now), and could define supplementary clusters if required, in a new spec field similar to the named cluster field in this PR.

In that sense I would amend the `run function` for all tests to receive an additional parameter for the supplementary clusters (which will be nil if none have been defined). This would eliminate the need to have either a multi-cluster type test or a normal single cluster type test and avoid some of the branching logic. We'd only then need to consider if there are additional supplementary clusters to do extra clean up work and considerations for saving clusters.

I might be missing something that forced the implementation to take the route in this PR, or having it the way I specified may not reduce the burden as much, but to me it feels more natural asking for additional clusters than switching to a new mode.

**Comment by herkolategan** [comment]
Another quick question, was there a TC run for this, and what do the artifacts look like for a multi-cluster run?

File: pkg/cmd/roachtest/test_runner.go [hunk 1]
```diff
@@ -1136,6 +1241,7 @@ func (r *testRunner) runTest(
 	stdout io.Writer,
 	l *logger.Logger,
 	github *githubIssues,
+	multiCluster cluster.MultiCluster,
```
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:1244]: `multiCluster` is unused which is probably a hint that the implementation isn't complete. E.g., `Monitor` is used only for a single cluster, whichever one happens to be last in `TestSpec.Clusters`; same for `MaybeExtendCluster`, as well as the whole tear down sequence, which deals with retrieving artifacts and triaging failure.

File: pkg/cmd/roachtest/test_runner.go [hunk 2]
```diff
@@ -760,20 +760,68 @@ func (r *testRunner) runWorker(
...
 
 		if c == nil {
 			// Create a new cluster if can't reuse or reuse attempt failed.
 			// N.B. non-reusable cluster would have been destroyed above.
 			wStatus.SetTest(nil /* test */, testToRun)
-			c, vmCreateOpts, clusterCreateErr = r.allocateOrAttachToCluster(
-				ctx, clusterFactory, clustersOpt, lopt,
-				testToRun.spec, arch, wStatus)
+
+			if testToRun.spec.IsMultiCluster() {
```
Comment (srosenberg) [inline pkg/cmd/roachtest/test_runner.go:770]: `testToRun` depends on `workPool.selectTest` which itself depends on `t.spec.Cluster.TotalCPUs()`. So, unless this multi-cluster spec is also a single-cluster spec, cpu quota is ignored.

**Comment by srosenberg** [comment]
> Thanks for having a look @DarrylWong and @herkolategan. Where does this leave us? We have a desire to test this functionality and no means to do so without either (a) a partial solution like the one propose here, or (b) a significant refactor. I have to presume that (b) isn't on your team's roadmap, so does it make sense to proceed with a partial solution for the time being?

I agree with the above comments. If we absolutely must shoehorn the support for multiple clusters, then we must eschew the use of `IsMultiCluster`.  For the most part, the (roachtest) framework is agnostic to how many clusters are created for a given test. I would suggest to start with changing the type of `TestSpec.Cluster` to `[]spec.ClusterSpec`. Majority of the existing tests use `MakeClusterSpec` which can be easily changed to return a singleton. Thus, we end up reusing the same (conditional) paths inside the test_runner. The type change also becomes a forcing function for addressing current implementation gaps like `workPool.selectTest` and `teardownTest`.

**Comment by ajstorm** [comment]
Thanks for the reviews @herkolategan and @srosenberg.  I'm tied up with a few higher priority items for the next few days, but will take another crack at this soon after.
